README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Modana

<!-- badges: start -->
<!-- badges: end -->

The goal `Modana` package is to implement a refinement of moderation
analysis with binary outcomes, as proposed by [Anto and Su
(2023)](https://journals.sagepub.com/doi/abs/10.1177/09622802231151206?journalCode=smma).
The function fits three models of interest: a direct model, an inverse
model, and a generalized estimating equation (GEE) model. The direct and
inverse models are fitted using the `glm` function to verify the
symmetry property of odds ratio/relative risk in moderation analysis for
the main treatment effect as well as the moderating effects. The GEE
model is fitted using the `geeglm` function and is used to estimate the
treatment effect accounting for within-cluster correlation. The `Modana`
package has three different built functions.

## Installation

You can install the development version of Modana like so:

``` r
install.packages("Modana_0.1.0.tar.gz", repos = NULL)
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(Modana)
## basic example code
```

You can simulate some data to run moderation analysis

``` r
set.seed(1099)
b0 <- c(1, 1.2, 1, 1.5, -2, -.5, 1.2, 1.5)
a0 <- c(-1, 1.5, -1, 1.5)
#a0 <- c(-2, 2, 2, 0, 1)
b0 <- c(-1.5, 1.5, 1.5, -2, 2, -.5, -1, 1.5)
 datt <- sim_data(n = 100, b0, a0 = NULL, binary.Xs = FALSE,
                      sigma = 1, uniform = FALSE, c0 = 1,
                      link.function = "logistic", rho = 0.2,
                      observational = FALSE, trt.p = 0.5,
                      interaction = 1:2, details = FALSE)
#> There are a total of 4 covariates
head(datt)
#>   y trt         x1         x2         x3         x4
#> 1 0   0 -1.0272485  1.5804971 -0.1492128  0.2379859
#> 2 1   0 -2.3314105 -1.6163845  1.9313538  1.8053311
#> 3 0   1  0.4191473 -0.7286691 -0.5131350  1.8232865
#> 4 0   0 -1.8775643  1.0827282  1.3265782 -1.1509396
#> 5 1   0  0.5414926 -0.1900738  0.4512489  0.4825833
#> 6 1   0 -0.6896955  0.6951233  1.9000859 -0.3240436
```

The `refinedmod()` function implements refined moderation analysis to
estimate both treatment effect and moderating effects associated with
two logistic regression models with binary outcomes/treatment via
generalized estimating equations (GEE). The function first performs a
role swap algorithm on the original data to obtain a clustered data of
size `2` (i.e., swapping the roles of the response variable and
treatment variable columns in the original data data and then stacking
the resultant data on top of the original data).

``` r
getres <- refinedmod(formula = y ~ trt + x1 + x2 + x3 + x4,
                     detail = TRUE, y = "y",
                    trt = "trt", data = datt,
                     effmod = c("x1", "x2"),
                      corstr = "independence")
#> The direct modeling fitting: 
#> =================================================================
#> 
#> Call:
#> glm(formula = form.direct, family = binomial(link = "logit"), 
#>     data = .SD)
#> 
#> Deviance Residuals: 
#>      Min        1Q    Median        3Q       Max  
#> -2.12526  -0.54951  -0.08492   0.53698   2.68620  
#> 
#> Coefficients:
#>             Estimate Std. Error z value Pr(>|z|)    
#> (Intercept)  -2.1411     0.6469  -3.310 0.000934 ***
#> trt           1.9063     0.7249   2.630 0.008545 ** 
#> x1            1.6592     0.6746   2.459 0.013921 *  
#> x2           -2.3898     0.6830  -3.499 0.000467 ***
#> x3            2.4399     0.5453   4.474 7.67e-06 ***
#> x4           -0.3810     0.2994  -1.273 0.203182    
#> trt:x1       -1.6094     0.7477  -2.153 0.031357 *  
#> trt:x2        1.4938     0.7740   1.930 0.053598 .  
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
#> 
#> (Dispersion parameter for binomial family taken to be 1)
#> 
#>     Null deviance: 136.058  on 99  degrees of freedom
#> Residual deviance:  71.649  on 92  degrees of freedom
#> AIC: 87.649
#> 
#> Number of Fisher Scoring iterations: 6
#> 
#> The inverse modeling fitting: 
#> =================================================================
#> 
#> Call:
#> glm(formula = form.inverse, family = binomial(link = "logit"), 
#>     data = .SD)
#> 
#> Deviance Residuals: 
#>     Min       1Q   Median       3Q      Max  
#> -1.8781  -0.9950   0.4503   1.0532   1.8161  
#> 
#> Coefficients:
#>             Estimate Std. Error z value Pr(>|z|)  
#> (Intercept)  -0.3216     0.3075  -1.046   0.2957  
#> y             1.6005     0.6344   2.523   0.0116 *
#> x1            0.1812     0.2571   0.705   0.4810  
#> x2           -0.2929     0.3610  -0.811   0.4172  
#> x3           -0.2798     0.3029  -0.924   0.3556  
#> x4            0.4367     0.2297   1.901   0.0573 .
#> y:x1         -0.6432     0.4373  -1.471   0.1414  
#> y:x2          1.2298     0.5382   2.285   0.0223 *
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
#> 
#> (Dispersion parameter for binomial family taken to be 1)
#> 
#>     Null deviance: 138.27  on 99  degrees of freedom
#> Residual deviance: 120.78  on 92  degrees of freedom
#> AIC: 136.78
#> 
#> Number of Fisher Scoring iterations: 4
#> 
#> The GEE modeling fitting: 
#> =================================================================
#>               Estimate Std. Error   z value     Pr(>|z|)
#> (Intercept) -1.9882558  0.6313724 -3.149102 1.637732e-03
#> trt          1.7420214  0.6742747  2.583548 9.778975e-03
#> x1           1.1263252  0.3902621  2.886073 3.900812e-03
#> x2          -2.1327750  0.4347750 -4.905468 9.320477e-07
#> x3           2.2584618  0.4912020  4.597827 4.269210e-06
#> x4          -0.4006063  0.2991221 -1.339273 1.804817e-01
#> z            1.6338187  0.5822168  2.806203 5.012902e-03
#> x1:z        -0.8611383  0.3730072 -2.308637 2.096371e-02
#> x2:z         1.8597407  0.4994187  3.723811 1.962380e-04
#> x3:z        -2.5671182  0.6832798 -3.757053 1.719264e-04
#> x4:z         0.8363693  0.3954295  2.115091 3.442221e-02
#> trt:x1      -0.9294065  0.4424454 -2.100613 3.567497e-02
#> trt:x2       1.2861794  0.5361274  2.399018 1.643911e-02
names(getres)
#> [1] "Directmodel"       "Inversemodel"      "Geerefinedmodel"  
#> [4] "out.combined"      "comparisonresults"
```

You can call each elements of listed models

Mostly importantly, we can print the results of the refined estimation.

``` r
print(getres$Geerefinedmodel)
#>               Estimate Std. Error   z value     Pr(>|z|)
#> (Intercept) -1.9882558  0.6313724 -3.149102 1.637732e-03
#> trt          1.7420214  0.6742747  2.583548 9.778975e-03
#> x1           1.1263252  0.3902621  2.886073 3.900812e-03
#> x2          -2.1327750  0.4347750 -4.905468 9.320477e-07
#> x3           2.2584618  0.4912020  4.597827 4.269210e-06
#> x4          -0.4006063  0.2991221 -1.339273 1.804817e-01
#> z            1.6338187  0.5822168  2.806203 5.012902e-03
#> x1:z        -0.8611383  0.3730072 -2.308637 2.096371e-02
#> x2:z         1.8597407  0.4994187  3.723811 1.962380e-04
#> x3:z        -2.5671182  0.6832798 -3.757053 1.719264e-04
#> x4:z         0.8363693  0.3954295  2.115091 3.442221e-02
#> trt:x1      -0.9294065  0.4424454 -2.100613 3.567497e-02
#> trt:x2       1.2861794  0.5361274  2.399018 1.643911e-02
```

Example usage on a real-world data

``` r
#install.packages("CERFIT")
library(CERFIT)
data(warts, package =  "CERFIT")
dat <- warts
#data wrangling
dat$Type <- ifelse(dat$Type==3, 1, 0)
dat$response <- dat$Result_of_Treatment
```

``` r
res <- refinedmod(formula = response ~ treatment + age + Time + Type,
                     detail = FALSE, y = "response",
                    trt = "treatment", data = dat,
                     effmod = c("age", "Type"),
                      corstr = "independence")
print(res$comparisonresults)
#> $Direct.model
#> 
#> Call:
#> glm(formula = form.direct, family = binomial(link = "logit"), 
#>     data = .SD)
#> 
#> Deviance Residuals: 
#>     Min       1Q   Median       3Q      Max  
#> -3.4483  -0.2552   0.2177   0.5278   1.8712  
#> 
#> Coefficients:
#>                Estimate Std. Error z value Pr(>|z|)    
#> (Intercept)     7.43389    1.44174   5.156 2.52e-07 ***
#> treatment       0.59584    1.31994   0.451   0.6517    
#> age            -0.03101    0.02545  -1.219   0.2230    
#> Time           -0.58627    0.11036  -5.312 1.08e-07 ***
#> Type           -0.54809    0.87790  -0.624   0.5324    
#> treatment:age  -0.05274    0.04053  -1.301   0.1931    
#> treatment:Type -2.84916    1.43163  -1.990   0.0466 *  
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
#> 
#> (Dispersion parameter for binomial family taken to be 1)
#> 
#>     Null deviance: 230.51  on 179  degrees of freedom
#> Residual deviance: 119.85  on 173  degrees of freedom
#> AIC: 133.85
#> 
#> Number of Fisher Scoring iterations: 6
#> 
#> 
#> $Inverse.model
#> 
#> Call:
#> glm(formula = form.inverse, family = binomial(link = "logit"), 
#>     data = .SD)
#> 
#> Deviance Residuals: 
#>     Min       1Q   Median       3Q      Max  
#> -2.2753  -1.0578   0.1139   1.0293   1.9538  
#> 
#> Coefficients:
#>               Estimate Std. Error z value Pr(>|z|)    
#> (Intercept)    1.33013    1.12689   1.180 0.237861    
#> response       1.12250    1.10380   1.017 0.309181    
#> age           -0.00712    0.02312  -0.308 0.758123    
#> Time          -0.09123    0.06679  -1.366 0.171959    
#> Type           1.87854    0.70810   2.653 0.007980 ** 
#> response:age  -0.07198    0.03264  -2.205 0.027433 *  
#> response:Type -3.63225    0.95771  -3.793 0.000149 ***
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
#> 
#> (Dispersion parameter for binomial family taken to be 1)
#> 
#>     Null deviance: 249.53  on 179  degrees of freedom
#> Residual deviance: 202.03  on 173  degrees of freedom
#> AIC: 216.03
#> 
#> Number of Fisher Scoring iterations: 4
#> 
#> 
#> $Combinedgee.model
#>                   Estimate Std. Error    z value     Pr(>|z|)
#> (Intercept)     7.42036779 2.12457045  3.4926438 4.782640e-04
#> treatment       0.94854255 1.09332337  0.8675773 3.856258e-01
#> age            -0.02664027 0.02350883 -1.1332029 2.571291e-01
#> Time           -0.60056451 0.16830032 -3.5684098 3.591544e-04
#> Type           -0.35357228 0.91818191 -0.3850787 7.001791e-01
#> z              -6.07463142 1.39817721 -4.3446792 1.394794e-05
#> age:z           0.01717663 0.02031301  0.8455974 3.977774e-01
#> Time:z          0.51570766 0.13299353  3.8776897 1.054531e-04
#> Type:z          2.13458124 0.65420080  3.2628839 1.102847e-03
#> treatment:age  -0.06485767 0.03032095 -2.1390378 3.243260e-02
#> treatment:Type -3.40781554 0.91201807 -3.7365658 1.865506e-04
```


NEWS.md
# Modana 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(Modana)

test_check("Modana")


tests/testthat/testing.R
test_that("Testing the number of objects in my output", {
  a0 <- c(-2, 2, 2, 0, 1)
  b0 <- c(-1, 0.5, 0.5, -1, .5, -0.5, 1, 0, .5)
  source("~/Modana/Modana/R/moderation.R") #<- [GEORGE]: Is this OK?
  datt <- sim_data(n = 100, b0, a0 = NULL, binary.Xs = FALSE,
                   sigma = 1, uniform = FALSE, c0 = 1,
                   link.function = "logistic", rho = 0.2,
                   observational = FALSE, trt.p = 0.5,
                   interaction = 1:2, details = FALSE)
  getres <- refinedmod(formula = y ~ trt + x1 + x2 + x3,
                       detail = TRUE, y = "y", trt = "trt",
                       data = datt, effmod = c("x1", "x2"),
                       corstr = "independence")
  expect_equal(length(getres), 5)
})



R/moderation.R
#' @title Refined moderation Analysis with binary outcomes
#' @description This package employs the symmetry property concerning the ratio of odds ratios, which suggests that heterogeneous treatment effects could
#' be equivalently estimated via a role exchange between the outcome and treatment variable in logistic regression
#' models. Refined inference on moderating effects is obtained by rearranging data and combing two models into
#' one via a generalized estimating equation (GEE) approach.
#' @param formula an object of class "formula" (or one that can be coerced to that class): a symbolic description of the model to be fitted.
#' @param data an optional data frame, list or environment (or object coercible by as.data.frame to a data frame) containing the variables in the model.
#' @param y  name of a binary response column in data.
#' @param trt name of a binary treatment column in data.
#' @param effmod a vector of potential effect modifiers used in the data.
#' @param detail a logical argument specified to print summary results of the direct and inverse models
#' @param ...  Other arguments, which are passed to the functions in \code{\link[geepack:geeglm]{geeglm}}.
#' @details Three models of interest are fitted. The generalized linear models \code{\link[stats:glm]{glm}} (for the direct and inverse models) are fitted to verify the symmetry property of OR/RR in moderation analysis for the main treatment effect as well as the moderating effects.
#' The main idea of this computing trick is to rearrange data into a format that micmicks a clustered or longitudinal data. A clustered data of size `2` is obtained simply by swapping the roles of the response variable \code{y} and treatment variable \code{trt}
#' columns in the original data \code{data} and then stacking the resultant data on top of the original \code{data}. The dummy variable \code{z}  is introduced to distinguish the two glm models.
#' @param data original data frame containing the variables in the model.
#' @return An object of type \code{glm} and \code{geeglm}.
#'
#' @references \href{Anto E, Su X. Refined moderation analysis with binary outcomes in precision medicine research. Statistical Methods in Medical Research. 2023;0(0). doi:10.1177/09622802231151206}{Anto E, Su X. Refined moderation analysis with binary outcomes in precision medicine research. Statistical Methods in Medical Research. 2023;0(0). doi:10.1177/09622802231151206}
#'
#' @examples
#' set.seed(1001)
#' b0 <- c(-1.5, 1.5, 1.5, -2, 2, -.5, 2, 1.5)
#' a0 <- c(-1, 1.5, -1, 1.5)
#' datt <- sim_data(n = 100, b0, a0 = NULL, binary.Xs = FALSE,
#'                   sigma = 1, uniform = FALSE, c0 = 1,
#'                    link.function = "logistic", rho = 0.2,
#'                    observational = FALSE, trt.p = 0.5,
#'                    interaction = 1:2, details = FALSE)
#'
#' getres <- refinedmod(formula = y ~ trt + x1 + x2 + x3 + x4,
#'                      detail = TRUE, y = "y",
#'                     trt = "trt", data = datt,
#'                      effmod = c("x1", "x2"),
#'                       corstr = "independence")
#' print(getres$comparisonresults)
#' @export
refinedmod <- function(formula, y = "y", trt = "trt",
                       effmod = NULL, data, detail = FALSE, ...){
  dat0 <- swaparr(data, y, trt)
  dat0 <- data.frame(dat0[order(dat0$ID), ], row.names = NULL)
  ID = factor(dat0$ID)
  data0 <- data.table::setDT(dat0)
  data1 <- data.table::setDT(data)
  form <- formula
  vars <- all.vars(form)
  covar <- vars[!vars %in% c(y, trt)]
  covars <- c(trt, covar)
  covars.inv <- c(y, covar)
  if(is.null(effmod)){
    form.gee <- stats::as.formula(paste(y, "~",  trt, "+(",
                                 paste0(covar, sep = "", collapse = "+"),")*z"))
    form.inverse <- stats::as.formula(paste(trt, "~",
                                     paste0(covars.inv, sep = "", collapse = "+")))
    form.direct <- stats::as.formula(paste(y, "~",
                                    paste0(covars, sep = "", collapse = "+")))
  } else{
    form.gee <- stats::as.formula(paste(y, "~",  trt, "+(",
                                 paste0(covar, sep = "", collapse = "+"),")*z +(",
                                 paste0(effmod, sep = "", collapse = "+"),"):",trt))

    form.inverse <- stats::as.formula(paste(trt, "~",
                                     paste0(covars.inv, sep = "", collapse = "+"), "+(",
                                     paste0(effmod, sep = "", collapse = "+"),"):", y))

    form.direct <- stats::as.formula(paste(y, "~",
                                    paste0(covars, sep = "", collapse = "+"), "+(",
                                    paste0(effmod, sep = "", collapse = "+"),"):", trt))
  }
  #Fitting the models
  fit.gee <- data0[, geepack::geeglm(form.gee, family = binomial(link = "logit"),
                          data = .SD, id = ID, ...)]
  fit.direct <- data1[, glm(form.direct, family =  binomial(link = "logit"), data = .SD)]
  fit.inverse <- data1[, glm(form.inverse, family = binomial(link = "logit"), data = .SD)]
  # fit.gee <- geepack::geeglm(formula = formula0, data = dat0, id = ID,
  #                            family = binomial, ...)
  # fit.direct <- stats::glm(form.direct, data = data,
  #                          family = binomial(link = "logit"))
  # fit.inverse <- stats::glm(form.inverse, data = data,
  #                    family = binomial(link = "logit"))
  if (detail){
    cat("The direct modeling fitting: \n")
    cat("=================================================================\n")
    print(summary(fit.direct))
  }

  if (detail){
    cat("The inverse modeling fitting: \n")
    cat("=================================================================\n")
    print(summary(fit.inverse))
  }
  out.direct <- summary(fit.direct)$"coefficients"
  out.direct <- summary(fit.direct)$"coefficients"
  out.inverse <- summary(fit.inverse)$"coefficients"
  out.gee <- summary(fit.gee)$coefficients
  #names(result.gee)
  out.gee$Wald <- (out.gee$Estimate)/(out.gee$Std.err )
  out.gee$'Pr(>|W|)' <- 2*pnorm (-abs(out.gee$Wald ))

  names(out.gee)[names(out.gee) == "Wald"] <- "z value"
  names(out.gee)[names(out.gee) == "Pr(>|W|)"] <- "Pr(>|z|)"
  names(out.gee)[names(out.gee) == "Std.err"] <- "Std. Error"
  if (detail){
    cat("The GEE modeling fitting: \n")
    cat("=================================================================\n")
    print(out.gee)
  }
  alres <- list(Direct.model = summary(fit.direct),
                Inverse.model = summary(fit.inverse),
                Combinedgee.model = out.gee)
  structure(list(Directmodel  = fit.direct,
                 Inversemodel = fit.inverse,
                 Geerefinedmodel = out.gee,
                 out.combined = fit.gee,
                 comparisonresults = alres),
            class = "Refinedmoderation")
}

# summary.refinedmod <- function(object, ...){
#
# }


#' @rdname refinedmod
#' @param data original data frame containing the variables in the model.
#' @param y  name of a binary response column in data.
#' @param trt name of a binary treatment column in data.
#' @details swaps the columns specified by \code{y} and \code{trt}, appends the swapped data frame to the original data frame \code{data}, and adds two new columns \code{(group and ID)} to the merged data frame
#' @return  New data frame created by swapping selected columns \code{y} and \code{trt} from \code{data} using the sequence of indices constructed mimicking a longitudinal with 2 clusters.
#' @export

swaparr <- function(data, y = "y", trt = "trt")
{
  y <- if (class(y)=="character" & is.na(suppressWarnings(as.integer(y))))
    which(colnames(data)==y) else as.integer(y)
  trt <- if (class(trt)=="character" & is.na(suppressWarnings(as.integer(trt))))
    which(colnames(data)==trt) else as.integer(trt)

  if (!(1<=y & y<=length(data))) stop( "`y` represents invalid index!" )
  if (!(1<=trt & trt<=length(data))) stop("`trt` represents invalid index!")
  DF <- data[if(y==trt) 1:length(data)
             else c((if (min(y, trt)==1) c()
                     else 1:(min(y, trt)-1) ),
                    (if (min(y,trt)+1 == max(y,trt))
                      (min(y,trt)+1):(max(y,trt)-1)
                     else c( max(y,trt), (min(y,trt)+1):(max(y,trt)-1),
                             min(y,trt))), (if (max(y,trt)==length(data))
                               c() else (max(y,trt)+1):length(data)))]
  n <- nrow(data)
  names(DF) <- names(data)
  Dat <- data.frame(rbind(data, DF))
  Dat$z <- rep(c(0, 1), c(n, n))
  Dat$ID <- rep(1:n, 2)
  return(Dat)
}


#is.even <- function(x) x %% 2 == 0

#expit <- function(x) (tanh(x/2)+1)/2  	# CORRECT & VERIFIED
# expit0 <- function(x) exp(x)/(1+exp(x))  # LOGISTIC FUNCTION, INVERSE OF LOGIT FUNCTION
#' @rdname refinedmod
#' @details The function \code{sim_data()} generates simulated data for a randomized controlled trial or an observational study
#' @param n the number of observations to generate.
#' @param b0 a vector of coefficients for the linear predictor in the outcome model. The length of \code{b0} should be equal to the number of covariates plus one (for the intercept term). The first element of \code{b0} corresponds to the intercept, and the remaining elements correspond to the coefficients for the covariates which may include include interaction coefficients.
#' @param a0 a vector of coefficients for the linear predictor in the treatment assignment model. The length of \code{a0} should be equal to the number of covariates plus one (for the intercept term). The first element of a0 corresponds to the intercept, and the remaining elements correspond to the coefficients for the covariates. This argument is only used if `observational = TRUE`.
#' @param a0 a vector of coefficients for the linear predictor in the treatment assignment model. The length of \code{a0} should be equal to the number of covariates plus one (for the intercept term). The first element of a0 corresponds to the intercept, and the remaining elements correspond to the coefficients for the covariates. This argument is only used if `observational = TRUE`.
#' @param link.function a character string specifying the link function to use in the outcome model. The default is `logistic`, which corresponds to the logistic link function. The other option is `exp`, which corresponds to the log-linear link function.
#' @param sigma the standard deviation of the covariate distribution. The default value is `1`.
#' @param uniform a logical value indicating whether to transform the covariates to a uniform distribution on the interval `[0, c0]`. The default value is `FALSE`.
#' @param c0 the upper limit of the uniform distribution used to transform the covariates. The default value is 1.
#' @param binary.Xs a logical value indicating whether to change some columns of continuous covariates into binary variables. The default value is `FALSE`.
#' @param rho the correlation parameter used to generate the covariance matrix. The default value is `0.2`.
#' @param observational a logical value indicating whether to generate data for an observational study. The default value is FALSE.
#' @param trt.p the probability of receiving the treatment. This argument is only used if `observational = FALSE`. The default value is 0.5.
#' @param interaction a vector of indices indicating which covariates should be used to create interaction terms with the treatment variable. The length of interaction should be less than or equal to the number of covariates. The default value is `NULL`, which indicates no interaction terms should be used.
#' @param details a logical value indicating whether to print details about the generated data. The default value is FALSE.
#' @export
sim_data <- function(n = 100, b0, a0 = NULL, link.function = "logistic",
                     sigma=1, uniform=TRUE, c0 = 1, binary.Xs=FALSE,
                     rho = 0.2, observational = FALSE, trt.p = 0.5,
                     interaction = NULL, details = F) {

  # Check b0 length is even
  #if (!is.even(length(b0)) || length(b0) < 2) stop("Length of b0 should be an even number!")
  #p <- (length(b0) - 2) / 2
  # Number of covariates
  if (length(b0) < 2 || !is.numeric(b0)) stop("Invalid b0 argument")
  p <- (length(b0) - 2)
  if(!is.null(interaction)){
    p <- length(b0) - (2 + length(interaction))
  }
  if (!is.null(a0) && (length(a0) != (p + 1) || !is.numeric(a0))) stop("Invalid a0 argument")
  if (!is.null(interaction) &&
      (length(interaction) > p || !is.numeric(interaction))) stop("Invalid interactions argument")
  cat("There are a total of", p, "covariates \n")
  # Generate covariance matrix
  S <- sapply(1:p, function(i){
    sapply(1:p, function(j){
      sigma^2*rho^(abs(i-j))})})

  # Generate X
  # X <- draw.d.variate.uniform(no.row = n, d = p, cov.mat = S)
  mu <- rep(0, p)
  X <- MASS::mvrnorm(n = n, mu = mu, Sigma = S)
  if (uniform) X <- stats::pnorm(X)*c0 	# PROB INTEGRAL TRANSFORM TO UNIFORM [0,c0]
  # CHANGE SOME COLUMNS OF X INTO BINARY
  if(binary.Xs){
    threshold <- ifelse(uniform, c0/2, 0);
    X <- (X > threshold)
  }
  # Treatment assignment
  if (observational) {
    if (is.null(a0)) stop("Observational study requires specifying a0.")
    if (length(a0) != (p + 1)) stop("Length of a0 must be consistent with b0 (i.e. length of a0 should be equal to ", paste(sum(p+1)))
    eta.trt <- as.vector(cbind(1, X) %*% a0)
    pi.trt <- (tanh(eta.trt/2)+1)/2
    trt <- stats::rbinom(n = n, size = 1, prob = pi.trt)
    if (details) print(cbind(p.treatment = mean(pi.trt), p.trt = mean(trt)))
  } else {
    trt <- stats::rbinom(n = n, size = 1, prob = trt.p)
  }

  # Model matrix
  X0 <- cbind(1, trt, X)
  if (!is.null(interaction)) {
    if (length(interaction) != length(unique(interaction))) stop("Duplicate entries in interactions argument")
    if (any(interaction < 1) || any(interaction > p)) stop("Invalid entries in interactions argument")
    if (length(interaction) > p) stop("Number of interaction terms should be less than or equal to the number of covariates.", "In your case,", paste("the number of covariates is ", p, "which is less than the number of interaction terms specified which is", length(interaction), ". Consider setting the length of the vector for the `interaction` argument to at most", p))
    X_int <- X[, c(interaction), drop = FALSE]
    X0 <- cbind(X0, X_int*trt)
  }

  # Model response
  eta <- as.vector(X0 %*% b0)
  if (details) print(cbind(eta.mean = mean(eta), eta.min = max(eta),
                           prop.positive = sum(eta > 0)/length(eta)))
  if (link.function == "exp") {
    pi0 <- exp(eta * (eta <= 0))  # LOGLINEAR
  } else {
    pi0 <-  (tanh(eta/2)+1)/2 #expit(eta)  # LOGISTIC
  }
  y <- stats::rbinom(n, size = 1, prob = as.vector(pi0))

  # Return data
  dat <- data.frame(cbind(y, trt, X))
  colnames(dat) <- c("y", "trt", paste0("x", 1:p))
  if (!is.null(interaction)) {
    datint <- cbind(dat, X_int*trt)
    colnames(datint)[(p+3):(p+(2+length(interaction)))] <- paste0("z", 1:length(interaction))
    if(details) print(utils::head(datint))
  }
  dat
}


